grpc-file-transfer/golang/grpc_file_streaming/grpc_stream_io.go

```diff
@@ -0,0 +1,120 @@
+package grpc_file_streaming
+
+import (
+	"crypto/sha1"
+	"encoding/hex"
+	"github.com/kurtosis-tech/stacktrace"
+	"github.com/sirupsen/logrus"
+	"io"
+	"math"
+)
+
+const (
+	// Kurtosis imposes maximum file size of 100MB
+	kurtosisMaximumFileSizeLimit = 100 * 1024 * 1024
+
+	// Files are streamed in chunk of 3MB. This is important to keep this below 4MB as GRPC has hard limit of 4MB
+	// per message
+	chunkSize = 3 * 1024 * 1024
+)
+
+// sendMessagesToStream is a helper function that takes as input the payload to be sent, split it into fixed-size chunks
+// and send it via the stream using sendViaStreamFunc.
+// grpcMsgConstructor must be implemented by the user of this function. it should produce an actual proto message object
+// from the previousChunkHash and the byte array for this chunk.
+func sendMessagesToStream[DataChunkProtoMessage interface{}](
+	payloadNameForLogging string,
+	payload []byte,
+	sendViaStreamFunc func(msg interface{}) error,
+	grpcMsgConstructor func(previousChunkHash string, contentChunk []byte) (*DataChunkProtoMessage, error),
+) error {
+	if len(payload) > kurtosisMaximumFileSizeLimit {
+		return stacktrace.NewError("The files you are trying to send, which are now compressed, exceed or reach " +
+			"100mb. Please reduce the total file size.")
+	}
+
+	var previousChunkHash string
+	hasher := sha1.New()
+	chunkNumber := 0
+	totalChunksNumber := int(math.Ceil(float64(len(payload)) / chunkSize))
+
+	for blockIdx := 0; blockIdx < totalChunksNumber; blockIdx++ {
+		logrus.Debugf("Sending content for %s. Block number %d/%d", payloadNameForLogging, chunkNumber, totalChunksNumber-1)
+
+		var contentChunk []byte
+		payloadStartIdx := blockIdx * chunkSize
+		if blockIdx == totalChunksNumber-1 {
+			// last chunk - copy everything that's left in the payload array
+			contentChunk = make([]byte, len(payload)-payloadStartIdx)
```

Comment:
Oh yes good call. I changed this bit last minute because the linter flagged it, I should have changed the above as well. Sorry for the confusion here 👍